ccl/sqlproxyccl: TestDenylistUpdate failed

ccl/sqlproxyccl.TestDenylistUpdate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8139543?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8139543?buildTab=artifacts#/) on master @ [250238cd29102391dddbc8cc71380090c49ce509](https://github.com/cockroachdb/cockroach/commits/250238cd29102391dddbc8cc71380090c49ce509):


```
=== RUN   TestDenylistUpdate
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDenylistUpdate539761000
    test_log_scope.go:79: use -show-logs to present logs inline
    proxy_handler_test.go:738: -- test log scope end --
    proxy_handler_test.go:738: Leaked goroutine: goroutine 1620035 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
        Leaked goroutine: goroutine 1620358 [IO wait]:
        internal/poll.runtime_pollWait(0x7ff2d7f11748, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc015f2c800?, 0xc01805e690?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc015f2c800, {0xc01805e690, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc015f2c800, {0xc01805e690?, 0xc011c6fe48?, 0x4973d2?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc0080ab768, {0xc01805e690?, 0xc018208ae0?, 0xc018208ae0?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc0012658c0, {0x6b86038?, 0xc018208ae0})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
--- FAIL: TestDenylistUpdate (7.39s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-sessions @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDenylistUpdate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-22896

ccl/sqlproxyccl.TestDenylistUpdate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8367332?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8367332?buildTab=artifacts#/) on master @ [b6434c942b4811c088349c43a81a46a53fb9b6c0](https://github.com/cockroachdb/cockroach/commits/b6434c942b4811c088349c43a81a46a53fb9b6c0):


```
=== RUN   TestDenylistUpdate
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDenylistUpdate450814667
    test_log_scope.go:79: use -show-logs to present logs inline
    proxy_handler_test.go:738: -- test log scope end --
    proxy_handler_test.go:738: Leaked goroutine: goroutine 1599569 [IO wait]:
        internal/poll.runtime_pollWait(0x7fe73095fb48, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc014e94200?, 0xc015ea80c0?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc014e94200, {0xc015ea80c0, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc014e94200, {0xc015ea80c0?, 0xc0111a9e48?, 0x4a74f2?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc0093320d8, {0xc015ea80c0?, 0xc01658f560?, 0xc01658f560?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc009fa3ce0, {0x6ccdc98?, 0xc01658f560})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
        Leaked goroutine: goroutine 1599811 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
--- FAIL: TestDenylistUpdate (6.75s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDenylistUpdate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl.TestDenylistUpdate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8382232?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8382232?buildTab=artifacts#/) on master @ [d698e5eaf36bee19901beeff7db3425315402907](https://github.com/cockroachdb/cockroach/commits/d698e5eaf36bee19901beeff7db3425315402907):


```
=== RUN   TestDenylistUpdate
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDenylistUpdate2221752237
    test_log_scope.go:79: use -show-logs to present logs inline
    proxy_handler_test.go:738: -- test log scope end --
    proxy_handler_test.go:738: Leaked goroutine: goroutine 1703327 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
        Leaked goroutine: goroutine 1703354 [IO wait]:
        internal/poll.runtime_pollWait(0x7fda9aaae858, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc0168a4a80?, 0xc01672eb60?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc0168a4a80, {0xc01672eb60, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc0168a4a80, {0xc01672eb60?, 0xc01345fe48?, 0x4a74f2?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc0107ad028, {0xc01672eb60?, 0xc01689ff20?, 0xc01689ff20?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc0075e7b80, {0x6cb1538?, 0xc01689ff20})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
--- FAIL: TestDenylistUpdate (7.21s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDenylistUpdate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl.TestDenylistUpdate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8396111?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8396111?buildTab=artifacts#/) on master @ [2861a4ee05fac1cc13f87bf00f368a65ea8700dc](https://github.com/cockroachdb/cockroach/commits/2861a4ee05fac1cc13f87bf00f368a65ea8700dc):


```
=== RUN   TestDenylistUpdate
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDenylistUpdate2459092626
    test_log_scope.go:79: use -show-logs to present logs inline
    proxy_handler_test.go:738: -- test log scope end --
    proxy_handler_test.go:738: Leaked goroutine: goroutine 1635050 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
        Leaked goroutine: goroutine 1637092 [IO wait]:
        internal/poll.runtime_pollWait(0x7f011523ba98, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc016e54080?, 0xc01672f090?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc016e54080, {0xc01672f090, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc016e54080, {0xc01672f090?, 0xc00fd4de48?, 0x4a74f2?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc0167b20c0, {0xc01672f090?, 0xc016e49620?, 0xc016e49620?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc01441edc0, {0x6cd8038?, 0xc016e49620})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
--- FAIL: TestDenylistUpdate (7.80s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDenylistUpdate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl.TestDenylistUpdate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8404119?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8404119?buildTab=artifacts#/) on master @ [b21379bb56dd206e6f63cc7d07ca72e85db7a4c4](https://github.com/cockroachdb/cockroach/commits/b21379bb56dd206e6f63cc7d07ca72e85db7a4c4):


```
=== RUN   TestDenylistUpdate
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDenylistUpdate2688960493
    test_log_scope.go:79: use -show-logs to present logs inline
    proxy_handler_test.go:738: -- test log scope end --
    proxy_handler_test.go:738: Leaked goroutine: goroutine 1598894 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
        Leaked goroutine: goroutine 1598945 [IO wait]:
        internal/poll.runtime_pollWait(0x7f2a90827968, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc013fc7f80?, 0xc017825bc0?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc013fc7f80, {0xc017825bc0, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc013fc7f80, {0xc017825bc0?, 0xc00db45e48?, 0x4a74f2?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc0065d6cd8, {0xc017825bc0?, 0xc017da80c0?, 0xc017da80c0?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc003b2e160, {0x6cdb078?, 0xc017da80c0})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
--- FAIL: TestDenylistUpdate (7.73s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDenylistUpdate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl.TestDenylistUpdate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8410646?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8410646?buildTab=artifacts#/) on master @ [b21379bb56dd206e6f63cc7d07ca72e85db7a4c4](https://github.com/cockroachdb/cockroach/commits/b21379bb56dd206e6f63cc7d07ca72e85db7a4c4):


```
=== RUN   TestDenylistUpdate
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDenylistUpdate970251351
    test_log_scope.go:79: use -show-logs to present logs inline
    proxy_handler_test.go:738: -- test log scope end --
    proxy_handler_test.go:738: Leaked goroutine: goroutine 1664446 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
        Leaked goroutine: goroutine 1664475 [IO wait]:
        internal/poll.runtime_pollWait(0x7f1fbc59aa08, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc014d5fb00?, 0xc01725dbf0?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc014d5fb00, {0xc01725dbf0, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc014d5fb00, {0xc01725dbf0?, 0xc00fc93e48?, 0x4a74f2?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc015efb318, {0xc01725dbf0?, 0xc017700240?, 0xc017700240?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc00820c6e0, {0x6cdb078?, 0xc017700240})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
--- FAIL: TestDenylistUpdate (7.64s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDenylistUpdate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl.TestDenylistUpdate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8426510?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8426510?buildTab=artifacts#/) on master @ [dc974cd698364a4db5acb630162dd8b1856cfad6](https://github.com/cockroachdb/cockroach/commits/dc974cd698364a4db5acb630162dd8b1856cfad6):


```
=== RUN   TestDenylistUpdate
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDenylistUpdate3003610292
    test_log_scope.go:79: use -show-logs to present logs inline
    proxy_handler_test.go:738: -- test log scope end --
    proxy_handler_test.go:738: Leaked goroutine: goroutine 1606851 [IO wait]:
        internal/poll.runtime_pollWait(0x7fecc069ec28, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc016148780?, 0xc017141a60?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc016148780, {0xc017141a60, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc016148780, {0xc017141a60?, 0xc0172afe48?, 0x4a74f2?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc0132d99b8, {0xc017141a60?, 0xc0172a1200?, 0xc0172a1200?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc00a149a20, {0x6ce2758?, 0xc0172a1200})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
        Leaked goroutine: goroutine 1606860 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
--- FAIL: TestDenylistUpdate (8.12s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDenylistUpdate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl.TestDenylistUpdate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8486857?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8486857?buildTab=artifacts#/) on master @ [78fe59d0536be35fcec68280142746fb91515551](https://github.com/cockroachdb/cockroach/commits/78fe59d0536be35fcec68280142746fb91515551):


```
=== RUN   TestDenylistUpdate
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDenylistUpdate1135164739
    test_log_scope.go:79: use -show-logs to present logs inline
    proxy_handler_test.go:738: -- test log scope end --
    proxy_handler_test.go:738: Leaked goroutine: goroutine 1601566 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
        Leaked goroutine: goroutine 1601867 [IO wait]:
        internal/poll.runtime_pollWait(0x7fd7f3750020, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc013b31c00?, 0xc016b387b0?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc013b31c00, {0xc016b387b0, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc013b31c00, {0xc016b387b0?, 0xc00d6efe48?, 0x4a74f2?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc012f00778, {0xc016b387b0?, 0xc016c41920?, 0xc016c41920?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc008d2b080, {0x6d72a78?, 0xc016c41920})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
--- FAIL: TestDenylistUpdate (7.20s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDenylistUpdate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl.TestDenylistUpdate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8501231?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8501231?buildTab=artifacts#/) on master @ [69dd453d0e61e258f402c5751de310405743cd18](https://github.com/cockroachdb/cockroach/commits/69dd453d0e61e258f402c5751de310405743cd18):


```
=== RUN   TestDenylistUpdate
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDenylistUpdate2285263358
    test_log_scope.go:79: use -show-logs to present logs inline
    proxy_handler_test.go:738: -- test log scope end --
    proxy_handler_test.go:738: Leaked goroutine: goroutine 1608683 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
        Leaked goroutine: goroutine 1610180 [IO wait]:
        internal/poll.runtime_pollWait(0x7fdf16f5fe18, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc016318f00?, 0xc0171a76f0?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc016318f00, {0xc0171a76f0, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc016318f00, {0xc0171a76f0?, 0xc0111ade48?, 0x4a74f2?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc00705db80, {0xc0171a76f0?, 0xc01786b0e0?, 0xc01786b0e0?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc01546eb00, {0x6d7b378?, 0xc01786b0e0})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
--- FAIL: TestDenylistUpdate (8.07s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDenylistUpdate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/sqlproxyccl/proxy_handler_test.go
+++ b/pkg/ccl/sqlproxyccl/proxy_handler_test.go
@@ -662,10 +662,15 @@ func TestDenylistUpdate(t *testing.T) {
 	defer log.Scope(t).Close(t)
 
 	ctx := context.Background()
-	te := newTester()
-	defer te.Close()
 
+	// Create an empty denylist file.
 	denyList, err := os.CreateTemp("", "*_denylist.yml")
+	require.NoError(t, err)
+	defer func() { _ = os.Remove(denyList.Name()) }()
+	dlf := denylist.File{Seq: 0}
+	bytes, err := dlf.Serialize()
+	require.NoError(t, err)
+	_, err = denyList.Write(bytes)
 	require.NoError(t, err)
 
 	sql, sqlDB, _ := serverutils.StartServer(t,
@@ -689,52 +694,77 @@ func TestDenylistUpdate(t *testing.T) {
 	proxyOutgoingTLSConfig := outgoingTLSConfig.Clone()
 	proxyOutgoingTLSConfig.InsecureSkipVerify = true
 
-	// We wish the proxy to work even without providing a valid TLS client cert to the SQL server.
+	// We wish the proxy to work even without providing a valid TLS client cert
+	// to the SQL server.
 	proxyOutgoingTLSConfig.Certificates = nil
+
+	// Register one SQL pod in the directory server.
+	tenantID := serverutils.TestTenantID()
+	tds := tenantdirsvr.NewTestStaticDirectoryServer(sql.Stopper(), nil /* timeSource */)
+	tds.CreateTenant(tenantID, "tenant-cluster")
+	tds.AddPod(tenantID, &tenant.Pod{
+		TenantID:       tenantID.ToUint64(),
+		Addr:           sql.ServingSQLAddr(),
+		State:          tenant.RUNNING,
+		StateTimestamp: timeutil.Now(),
+	})
+	require.NoError(t, tds.Start(ctx))
 
 	originalBackendDial := BackendDial
 	defer testutils.TestingHook(&BackendDial, func(
 		msg *pgproto3.StartupMessage, outgoingAddress string, tlsConfig *tls.Config,
 	) (net.Conn, error) {
-		time.AfterFunc(100*time.Millisecond, func() {
-			dlf := denylist.File{
-				Denylist: []*denylist.DenyEntry{
-					{
-						Entity:     denylist.DenyEntity{Type: denylist.IPAddrType, Item: "127.0.0.1"},
-						Expiration: timeutil.Now().Add(time.Minute),
-						Reason:     "test-denied",
-					},
-				},
-			}
-
-			bytes, err := dlf.Serialize()
-			require.NoError(t, err)
-			_, err = denyList.Write(bytes)
-			require.NoError(t, err)
-		})
 		return originalBackendDial(msg, sql.ServingSQLAddr(), proxyOutgoingTLSConfig)
 	})()
 
-	s, addr, _ := newSecureProxyServer(ctx, t, sql.Stopper(), &ProxyOptions{
+	opts := &ProxyOptions{
 		Denylist:           denyList.Name(),
 		PollConfigInterval: 10 * time.Millisecond,
-	})
-	defer func() { _ = os.Remove(denyList.Name()) }()
+	}
+	opts.testingKnobs.directoryServer = tds
+	s, addr, _ := newSecureProxyServer(ctx, t, sql.Stopper(), opts)
 
-	url := fmt.Sprintf("postgres://testuser:foo123@%s/defaultdb_29?sslmode=require&options=--cluster=tenant-cluster-28&sslmode=require", addr)
-	te.TestConnect(ctx, t, url, func(conn *pgx.Conn) {
-		require.Eventuallyf(
-			t,
-			func() bool {
-				_, err = conn.Exec(context.Background(), "SELECT 1")
-				return err != nil
-			},
-			time.Second, 5*time.Millisecond,
-			"Expected the connection to eventually fail",
-		)
-		require.Regexp(t, "unexpected EOF|connection reset by peer", err.Error())
-		require.Equal(t, int64(1), s.metrics.ExpiredClientConnCount.Count())
-	})
+	// Establish a connection.
+	url := fmt.Sprintf("postgres://testuser:foo123@%s/defaultdb?sslmode=require&options=--cluster=tenant-cluster-%s&sslmode=require", addr, tenantID)
+	db, err := gosql.Open("postgres", url)
+	db.SetMaxOpenConns(1)
+	defer db.Close()
+	require.NoError(t, err)
+
+	// Use a single connection so that we don't reopen when the connection
+	// is closed.
+	conn, err := db.Conn(ctx)
+	require.NoError(t, err)
+
+	_, err = conn.ExecContext(ctx, "SELECT 1")
+	require.NoError(t, err)
+
+	// Once connection has been established, attempt to update denylist.
+	dlf.Seq++
+	dlf.Denylist = []*denylist.DenyEntry{
+		{
+			Entity:     denylist.DenyEntity{Type: denylist.IPAddrType, Item: "127.0.0.1"},
+			Expiration: timeutil.Now().Add(time.Minute),
+			Reason:     "test-denied",
+		},
+	}
+	bytes, err = dlf.Serialize()
+	require.NoError(t, err)
+	_, err = denyList.Write(bytes)
+	require.NoError(t, err)
+
+	// Subsequent Exec calls will eventually fail.
+	require.Eventuallyf(
+		t,
+		func() bool {
+			_, err = conn.ExecContext(ctx, "SELECT 1")
+			return err != nil
+		},
+		time.Second, 5*time.Millisecond,
+		"Expected the connection to eventually fail",
+	)
+	require.Regexp(t, "closed|bad connection", err.Error())
+	require.Equal(t, int64(1), s.metrics.ExpiredClientConnCount.Count())
 }
 
 func TestDirectoryConnect(t *testing.T) {

